Reworks entity cache service and memcache, adds tests

hypeJunction opened the pull request:
Prepping the foundation for user table tests.

mrclay committed:
chore(cache): reworks ElggMemcache and its core usage

Properly loads extra query selects into memcache-loaded entities.
Properly updates the last_action/prev_last_action on current user.
No longer caches entity volatile data.
Caches subtype table data.
Adds support for the other `memcached` extension.
Adds a null cache to simplify logic.
Removes usage around metadata since it doesn't save any values.
Makes tests more resilient to failures.
Allows service provider test to support nullable types.

Fixes #7643
Fixes #7644
Fixes #7533

mrclay committed:
chore(memcache): reduce useless churn in memcache entity storage

As we have to load full DB rows anyway, we do nothing but churn memcache
trying to load and save every entity in fetched lists. This refocuses the
entity strategy on single get_entity() calls.

mrclay committed:
chore(session): don't write DB/cache if last_action won't change

hypeJunction committed:
chore(entities): refactor users and entity tables

Refactors database queries in entity and user tables
Moves memcache logic into entity cache service
Improves cache flushing logic

hypeJunction commented on engine/lib/cache.php:
```diff
@@ -241,6 +241,19 @@ function _elgg_is_cache_symlinked() {
+ *
+ * @return void
+ * @access private
+ * @deprecated 2.3
```
Not yet :)

hypeJunction commented on engine/tests/ElggUserTest.php:
```diff
@@ -177,6 +179,9 @@ public function testElggUserRemoveAdmin() {
 		$this->assertTrue($this->user->removeAdmin());
 
+		// need to save user to propagate attributes to the database
+		$this->user->save();
```
@mrclay Any clue how this test was passing all this time?

hypeJunction replied:
Never mind. I broke something. I thought it was just setting the attribute without saving the entity.

hypeJunction replied:
There was a bug in `ElggEntity::delete()`. Cache was invalidated, but it was being repopulated in one of the consequent delete operations (not sure which one - metadata, relationships or other). I have moved cache invalidation closer to the DB delete query.

hypeJunction commented on engine/tests/ElggUserTest.php:
```diff
@@ -131,7 +131,9 @@ public function testElggUserNameCache() {
 		$user = get_user_by_username($name);
-		$user->delete();
+
+		$this->assertTrue($user->delete());
```
@mrclay Can't figure out why this one is failing

hypeJunction replied:
There was a bug in `ElggEntity::delete()`. Cache was invalidated, but it was being repopulated in one of the consequent delete operations (not sure which one - metadata, relationships or other). I have moved cache invalidation closer to the DB delete query.

mrclay commented:
This will make https://github.com/Elgg/Elgg/pull/9080 even more nightmarish to rebase. Could you take a look at that first? Ideally we could add Travis memcache tests, try to get #9080 in then move reads into this. It may be beyond saving anyway.

hypeJunction commented:
We shouldn't let anything sit for more than a week. At this pace of refactoring everything goes stale. I tried to salvage my notifications PR that's how I ended up reworking this. Waste of time.

hypeJunction commented:
Maybe you should get in #10091 before I continue with this. 
I can look at the memcache PR in detail and port over necessary changes when I get to it.

hypeJunction committed:
chore(tests): adds EntityCache tests

hypeJunction commented:
Added entity cache tests

hypeJunction commented:
@mrclay I think I have saved #9080. I will hold off with user table testing until we merge this in.

hypeJunction commented:
Something is wonky with permissions following #10103

hypeJunction commented:
I have no clue what's breaking the tests that were passing before I added #9080. I can't spend any more time debugging.

mrclay commented on engine/classes/Elgg/Cache/EntityCache.php:
```diff
@@ -65,6 +72,20 @@ public function get($guid) {
+	 * @return \ElggUser|false
+	 */
+	public function getByUsername($username) {
+		$cache = array_flip($this->username_cache);
```
Use usernames as keys?

hypeJunction replied:
We would have to flip to remove from cache anyway...

mrclay replied:
Then use array_search() for this.

hypeJunction replied:
Added in 1.8.0.1 it seems https://github.com/Elgg/Elgg/commit/c80f32c614e48e237bcc3b72b780badf17d99dca

hypeJunction replied:
Good point

mrclay commented:
I can dig more into this later. Exciting!

hypeJunction commented:
Why do we maintain two sets of caches? Is there really a use case for keeping memcache separate from entity cache service? Looking at the code we just have two duplicate queries everywhere for set/remove operations.

hypeJunction commented:
@mrclay when you look at it. The tests were failing due to lacking permissions. I thought userCapabilities needed to be reinstantiated with the new mock entities table, but that had no effect. We are loosing logged in user somewhere it seems.

mrclay commented:
> Why do we maintain two sets of caches?

The in-memory cache is essential to avoid tons of duplicate queries or repeating construction costs (if we relied just on the query cache). Memcache, on the other hand, is severely limited since we at least must run an access query for every entity. I'm not sure the in-memory query cache is even needed anymore.

mrclay commented on engine/classes/Elgg/Database/UsersTable.php:
```diff
@@ -494,64 +516,96 @@ function setValidationStatus($user_guid, $status, $method = '') {
-		return false;
+	public function getValidationStatus($user_guid) {
+		$user = get_entity($user_guid);
+		return $user->validated;
```
```php
if (!$user || !isset($user->validated)) {
  return null
}
return (bool) $user->validated;
```

mrclay commented:
Up to the tests this looks great, will dig into that tomorrow.

hypeJunction committed:
chore(tests): rework entity database queries and caching

Audits in-memory and memcache caching strategy for entities
Optimizes DB queries in entity and user tables to minimize
required read operations (by standardizing queries and relying
on DB query cache)
Modifies mocking strategy in PHPUnit tests and increases test
coverage
Fixes various bugs revealed by tests
Changes in entity and user last action timestamps are now consistently
propagated to entity caches and database
Deprecates update_entity_last_time() and adds ElggEntity::updateLastAction()
for private use instead

hypeJunction commented:
@mrclay I was able to figure out the problem with tests. We were instantiating a new session for each test in a test case, but mock services were served from a singleton, so they never got the new session value. I have reworked the mocking part a bit, so we now can use a clean slate for each test in a case.

I have added last action update tests, and while testing discovered bugs, which are fixed now.

hypeJunction commented on engine/lib/entities.php:
```diff
@@ -741,11 +741,15 @@ function elgg_instanceof($entity, $type = null, $subtype = null, $class = null)
  */
 function update_entity_last_action($guid, $posted = null) {
-	return _elgg_services()->entityTable->updateLastAction($guid, $posted);
+	$entity = get_entity($guid);
```
We need ignored access here, I suppose

mrclay replied:
How long has this been private? We could just get rid of it.

hypeJunction replied:
Added in 1.8.0.1 it seems c80f32c

hypeJunction replied:
Let's just deprecate for now

hypeJunction commented on engine/classes/Elgg/Database/EntityTable.php:
```diff
@@ -217,13 +296,29 @@ public function get($guid, $type = '') {
+			$entity = $memcache->load($guid);
+			if ($entity instanceof ElggEntity) {
+				// until ACLs in memcache, DB query is required to determine access
+				if (!$this->getRow($guid)) {
```
Let's remove this and check for ignored access || has_access_to_entity below. We need to make sure entity is accesible even if it comes from in-memory cache.

hypeJunction commented:
This increases test coverage by another 4%.

hypeJunction commented on engine/classes/Elgg/Database/EntityTable.php:
```diff
@@ -217,13 +296,32 @@ public function get($guid, $type = '') {
+		}
+
+		// Validate accessibility
+		if ($entity && !elgg_get_ignore_access() && !has_access_to_entity($entity)) {
```
See the test at https://github.com/Elgg/Elgg/pull/10099/files#diff-b8a1d091cc488c59818e105a3cfd063cR116 that proves my point

hypeJunction replied:
But kind of defeats the purpose of a cache... I guess getRow() would have been cached in DB queries at least.

hypeJunction replied:
Unless we invalidate in-memory cache when ignored access is toggled, and user is logged in/out, and only do this for memcache. But to be honest, I am quite disillusioned with memcache for entities - doesn't really add much gain for increased complexity.

mrclay replied:
For in-memory we *were* refusing to cache when access was off and clearing cache on session change. That logic may have got broken along the way.

mrclay replied:
Agree that entities-in-memcache seems overly complex. We can simplify by only caching public ones, but even that's not safe considering the access query is hookable.

hypeJunction replied:
I need to switch to something else for now.  I start missing obvious things. Feel free to make PRs. I guess this will need to be squashed somehow.

mrclay replied:
Users would be good to have in there, since every logged-in boot has to fetch one and there's lots of user prefetching. The user prefetcher could try memcache first...

hypeJunction replied:
When an instance of `ElggEntity` is cached in memcached, does it also store all the metadata? I mean `$entity->foo = $bar` in theory should invalidate memcache.

mrclay replied:
Metadata is all separate, has to be because it's access controlled :(

hypeJunction replied:
> For in-memory we were refusing to cache when access was off and clearing cache on session change. That logic may have got broken along the way.

That was true for caching, but false for session change invalidation. I have made the necessary changes and added tests. On their way.

hypeJunction commented:
argh, @mrclay any clue how this query is even possible?
```sql
select * from elgg_t_i_entities where guid = :guid and ((access_id in (2,1)) and (enabled = 'yes'))
````

I am debugging access.

hypeJunction commented:
Our access queries are out of control... Argh. Will need to dig into them later.

hypeJunction commented:
It was a f**** chain reaction. Everything is failing now. Can't be bothered with it any more for now. Mocking queries with our bizarre access controls is impossible. It's absolutely random.

hypeJunction commented:
Leaving access changes out of it. It's a nightmare. Not sure we will ever be able to get them tested statically.

mrclay commented:
Let it sit. I'm upgrading my OS and can jump in later

hypeJunction commented:
Yeah. I need to switch off. Gotten a bit OCD with it.

On Aug 25, 2016 9:37 PM, "Steve Clay" <notifications@github.com> wrote:

> Let it sit. I'm upgrading my OS and can jump in later
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Elgg/Elgg/pull/10099#issuecomment-242511026>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABJaSUmTiDAtcAfiEibAVH5sXvFqbPF7ks5qje7tgaJpZM4JoSvz>
> .
>

hypeJunction commented:
Beer and sleep do miracles. Had a few revelations this morning. All green now.

mrclay commented on engine/classes/Elgg/Database/UsersTable.php:
```diff
@@ -555,34 +554,39 @@ public function setLastAction($user_guid) {
-			});
-		}
+
+		$this->entity_cache->set($user);
```
Probably already in the cache, but doesn't hurt.

mrclay commented on engine/classes/Elgg/Database/UsersTable.php:
```diff
@@ -593,19 +597,23 @@ public function setLastLogin($user_guid) {
-				_elgg_invalidate_memcache_for_entity($user_guid);
-			});
-		}
+		$this->entity_cache->set($user);
```
Like above, very likely already cached.

hypeJunction commented on engine/classes/Elgg/Database/EntityTable.php:
```diff
@@ -247,17 +356,18 @@ public function get($guid, $type = '') {
+		$ia = $this->session->setIgnoreAccess(true);
+		$show_hidden = access_show_hidden_entities(true);
+
+		$result = $this->getRow($guid);
```
The only problem with this is someone adding `and` clauses to the access sql. Should we even be worrying about someone ignoring ignored access in the hook handler?

mrclay replied:
Good point, I'm not sure anyone thought of that. "Ignore access" to me means `(1=1)` no matter what the hook says.

hypeJunction replied:
Let's just fix the "bug" in 2.2

mrclay replied:
Could you make an issue for this if not exists already?

mrclay commented on engine/tests/phpunit/Elgg/Database/EntityTableTest.php:
```diff
@@ -0,0 +1,82 @@
+		";
+
+		$new_last_action = strtotime('+2 days', $time);
+		_elgg_services()->db->addQuerySpec([
```
We should be checking that these queries are actually called

hypeJunction replied:
Might not be so easy but we can try.

mrclay replied:
I have an easy idea here.

mrclay commented:
I think you should set this aside for a bit.

The framework for mocking DB queries is great and we should expand it. It allows tests to run production code, and if we change code, the mock queries are easy to keep in sync.

The reliance on custom mock components (e.g. `Elgg\Mocks\Database\EntityTable`) I'm very worried about. It allows tests to bypass production code, which produces a deceiving coverage metric. And they're getting super complex; I've been following this PR closely and don't understand a lot of [this stuff](https://github.com/Elgg/Elgg/pull/10099/commits/67522633c65a07cdebcbea8eee27ea54421f7225#diff-16f0bee8d89f64b36a21c8c80e45afcfR240), how are other devs going to refactor it?

I'd rather start accepting the improvements bit by bit without tests, than to take in tests I'm not confident we can maintain. Better to let tests grow stale; at least then we can see how hard it is to update them.

hypeJunction commented:
I can break down the entity table mock and add detailed comments. The idea is that when you save an entity there is a set of DB queries with predictable results, i.e. select an entity by guid. Given there are numerous combinations for the access query (entity may be requested for logged in/logged out with/without ignored access and with/without hidden access). The access bit is reverse engineered, so I test the access level against known access levels to see if when the query is run the logged in user actually has access to it before returning it with results.

In any case, if I can't get this in, than I will probably stop writing any more services. Notifications features that I wrote would need this to be tested, and they would require proper entity mocks. I am not pushing any more code without tests, and I will be blocking any new features without tests.

To be clear the tests here do not test actual entity/user tables, those would be tested separately similar to subtype tables. What's I am trying to do is create an intereface where we can use DB APIs for testing other services. DB tables would get their own tests with mocking done on per method basis.

hypeJunction commented:
Setting this aside would mean just closing it, because it will never get merged - it will grow stale in a week.
I will try to add entity table and user table tests to demonstrate what I have in mind. I hope I can get it done before my holiday next week.

mrclay commented:
> Setting this aside would mean just closing it, because it will never get merged

Not true. We rescue stale work all the time and it's usually easy, and when updating I usually find the solution could be improved/simplified. So I don't have a problem with my PRs sitting around. I *do* agree that we should adjust the bar of acceptance to match the team's capacity for reviewing.

I should be able to do a deeper look tonight. Yesterday's OS upgrade aftermath left me reduced mental bandwidth.

> create an interface where we can use DB APIs for testing other services. DB tables would get their own tests with mocking done on per method basis.

Provided its maintainable, sounds good.

> Notifications features that I wrote would need this to be tested

Have you considered the simpletest suite? Testing the real integration is critical anyway.

hypeJunction commented:
> Have you considered the simpletest suite? Testing the real integration is critical anyway.
Well, that's the thing. Most of the tests we write do not really need DB connection. We just access entity properties and then discard them. Notifications test would be best done statically. I also prefer static analysis as it results in cleaner API. You can as well test individual methods without the overhead of system hooks/events etc that yield unexpected results.

hypeJunction committed:
chore(tests): clean up entity table and metadata table mocks

hypeJunction commented:
Entity table mock should be a bit easier to read.

mrclay commented:
When memcache is on `ElggCoreUserTest::testElggUserConstructorByDbRow` still fails due to last_action (#7644), but that's OK for now.

mrclay commented on engine/classes/Elgg/Cache/EntityCache.php:
```diff
@@ -105,14 +129,19 @@ public function remove($guid) {
 		unset($this->entities[$guid]);
 		$GLOBALS['ENTITY_CACHE'] = $this->entities;
 
+		$username = array_search($guid, $this->username_cache);
```
$key instead of $username.

mrclay commented on engine/classes/Elgg/Database/EntityTable.php:
```diff
@@ -217,13 +300,31 @@ public function get($guid, $type = '') {
+				return elgg_instanceof($entity, $type) ? $entity : false;
 			}
-			return $new_entity;
+			$this->entity_cache->set($entity);
```
Sloppy that even entities pulled from entity_cache are re-saved there, but if it bugs me I can rework later.

mrclay replied:
Sloppy is bad choice of words. It assumes `entity_cache->set()` has no cost, which is fairly true for now at least.

mrclay commented:
**LGTM** Rebase to kill the conflict and merge this beast.
